### PR TITLE
Enable/Disabled worker jobs in frontend

### LINKED ISSUE #7485: Jobs that are not available for a particular datastore should already be disabled in the frontend
#7463 introduces a new field `jobsSupportedByAvailableWorkers` for the dataset json. This should be used to decide which jobs should be shown in the UI for a dataset

### PR DESCRIPTION
**Pull request by hotzenklotz**
This PR adds the frontend part to #7463, i.e. enabling/disabling different UI elements if a certain worker jobs is not supported by that WK instance.

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- Adjust the worker's supported job commands in the `workers` table in the Postgres DB
- Check if the corresponding UI feature reacts either by being disabled or hidden altogether

### TODOs:
- [x] CONVERT_TO_WKW
- [x] EXPORT_TIFF
- [x] RENDER_ANIMATION
- [x] COMPUTE_MESH_FILE
- [x] COMPUTE_SEGMENT_INDEX_FILE
- [x] FIND_LARGEST_SEGMENT_ID
- [x] INFER_NUCLEI
- [x] INFER_NEURONS
- [x] MATERIALIZE_VOLUME_ANNOTATION

### Issues:
- fixes #7485

------
(Please delete unneeded items, merge only when none are left open)
- [x] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-libs python client](https://github.com/scalableminds/webknossos-libs/tree/master/webknossos/webknossos/client) if relevant API parts change
- [ ] Removed dev-only changes like prints and application.conf edits
- [ ] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)
- [ ] Needs datastore update after deployment

### CODE REVIEW

**Comment by hotzenklotz** [comment]
@philippotto Can you please review the frontend changes. Thank you.

File: frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx [hunk 1]
```diff
@@ -681,13 +681,18 @@ class SegmentsView extends React.Component<Props, State> {
...
+        APIJobType.COMPUTE_MESH_FILE,
+      )
+    ) {
+      title = "Mesh computation jobs are not enabled for this WEBKNOSSOS instance.";
     } else if (this.props.activeUser == null) {
       title = "Please log in to precompute the meshes of this dataset.";
-    } else if (!this.props.dataset.jobsEnabled) {
+    } else if (!this.props.dataset.dataStore.jobsEnabled) {
       title =
-        "Meshes Computation is not supported for datasets that are not natively hosted on the server. Upload your dataset directly to weknossos.org to enable this feature.";
```
Comment (philippotto) [inline frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx:690]: lol

File: frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx [hunk 2]
```diff
@@ -681,13 +681,18 @@ class SegmentsView extends React.Component<Props, State> {
       };
     }
 
-    if (!features().jobsEnabled) {
-      title = "Computation jobs are not enabled for this WEBKNOSSOS instance.";
+    if (
+      !features().jobsEnabled ||
+      !this.props.dataset.dataStore.jobsSupportedByAvailableWorkers.includes(
```
Comment (philippotto) [inline frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx:686]: is it guaranteed that jobsSupportedByAvailableWorkers is empty if jobsEnabled is false? Then, we could remove the jobsEnabled check almost everywhere, right?

File: frontend/javascripts/oxalis/view/right-border-tabs/bounding_box_tab.tsx [hunk 3]
```diff
@@ -109,12 +109,14 @@ export default function BoundingBoxTab() {
...
             color={bb.color}
             name={bb.name}
-            isExportEnabled={dataset.jobsEnabled}
+            isExportEnabled={dataset.dataStore.jobsEnabled}
             isVisible={bb.isVisible}
             onBoundingChange={_.partial(handleBoundingBoxBoundingChange, bb.id)}
             onDelete={_.partial(deleteBoundingBox, bb.id)}
             onExport={
-              dataset.jobsEnabled ? _.partial(setSelectedBoundingBoxForExport, bb) : () => {}
+              dataset.dataStore.jobsEnabled
```
Comment (philippotto) [inline frontend/javascripts/oxalis/view/right-border-tabs/bounding_box_tab.tsx:117]: no check with jobsSupportedByAvailableWorkers?

File: frontend/javascripts/oxalis/view/action_bar_view.tsx [hunk 4]
```diff
@@ -218,13 +218,13 @@ class ActionBarView extends React.PureComponent<Props, State> {
   }
 
   renderStartTracingButton(): React.ReactNode {
-    // @ts-expect-error ts-migrate(2345) FIXME: Argument of type '(props: AsyncButtonProps) => Ele... Remove this comment to see the full error message
```
Comment (philippotto) [inline frontend/javascripts/oxalis/view/action_bar_view.tsx:221]: nice :+1:

File: frontend/javascripts/oxalis/view/action_bar_view.tsx [hunk 5]
```diff
@@ -250,11 +250,10 @@ class ActionBarView extends React.PureComponent<Props, State> {
...
     const isAIAnalysisEnabled = () => {
-      const activeUser = this.props.activeUser;
-      const jobsEnabled = features().jobsEnabled;
-      if (isViewMode) {
-        return jobsEnabled && activeUser != null && activeUser.isSuperUser;
-      }
+      const jobsEnabled =
+        features().jobsEnabled &&
+        (dataset.dataStore.jobsSupportedByAvailableWorkers.includes(APIJobType.INFER_NEURONS) ||
+          dataset.dataStore.jobsSupportedByAvailableWorkers.includes(APIJobType.INFER_NUCLEI));
```
Comment (philippotto) [inline frontend/javascripts/oxalis/view/action_bar_view.tsx:256]: It looks like isAIAnalysisEnabled is only used to call renderStartAIJobButton which then uses `setAIJobModalStateAction("neuron_inferral")`. So, no nuclei? Maybe I'm missing something.

File: frontend/javascripts/oxalis/view/action-bar/toolbar_view.tsx [hunk 6]
```diff
@@ -443,16 +452,17 @@ function AdditionalSkeletonModesButtons() {
...
-          style={NARROW_BUTTON_STYLE}
-          onClick={() => setShowMaterializeVolumeAnnotationModal(true)}
-          title="Materialize this merger mode annotation into a new dataset."
-        >
-          <ExportOutlined />
-        </ButtonComponent>
-      )}
-      {features().jobsEnabled && showMaterializeVolumeAnnotationModal && (
+      {isMergerModeEnabled &&
+        isMaterializeVolumeAnnotationEnabled(
```
Comment (philippotto) [inline frontend/javascripts/oxalis/view/action-bar/toolbar_view.tsx:456]: this is called as a function but it's a boolean?

**Review by philippotto — CHANGES_REQUESTED** [review]
mostly looks good, but see my few comments.

File: frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx (hunk 2 above)
Comment (philippotto) [reply frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx:686]: (I think, I'd actually prefer this)

Comment (hotzenklotz) [reply frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx:686]: I wasn't sure either. Looks like `jobsEnable == False`  leads to an empty array of supported jobs. So, I can apply your suggestion.

image

File: frontend/javascripts/oxalis/view/action_bar_view.tsx (hunk 5 above)
Comment (hotzenklotz) [reply frontend/javascripts/oxalis/view/action_bar_view.tsx:256]: Yes, the UI only let's you trigger neuron jobs ATM. Nucleii maybe reactivated in the future, but the job types enum still defines them so I included them for completeness. Hence an `OR` condition.

File: frontend/javascripts/oxalis/view/action-bar/toolbar_view.tsx (hunk 6 above)
Comment (hotzenklotz) [reply frontend/javascripts/oxalis/view/action-bar/toolbar_view.tsx:456]: Good catch. This is wrong

**Comment by hotzenklotz** [comment]
@philippotto I applied your review suggestions / feedback. Please have a look.

File: frontend/javascripts/admin/dataset/dataset_upload_view.tsx [hunk 7]
```diff
@@ -606,6 +603,22 @@ class DatasetUploadView extends React.Component<PropsWithFormAndRouter, State> {
...
 
+  isDatasetConversionEnabled = () => {
+    const uploadableDatastores = this.props.datastores.filter(
+      (datastore) => datastore.allowsUpload,
+    );
+    const hasOnlyOneDatastoreOrNone = uploadableDatastores.length <= 1;
+
+    const selectedDatastore = hasOnlyOneDatastoreOrNone
+      ? uploadableDatastores[0]
+      : this.getDatastoreForUrl(this.state.datastoreUrl);
```
Comment (philippotto) [inline frontend/javascripts/admin/dataset/dataset_upload_view.tsx:614]: `this.state.datastoreUrl` is only set once the user submits the form. The antd form logic is quite messy, unfortunately. Before we do the setState, the url lives in the antd form object and can be retrieved via `currentFormRef.getFieldValue("datastoreUrl")`.

My suggestion would be to use `<Form ... onValuesChange={...}>` and in that handler, always set this.state.datastoreUrl. That way, both vars are synchronized. Then, the setState in the handleSubmit code does not need to set the datastoreUrl anymore.

Comment (hotzenklotz) [reply frontend/javascripts/admin/dataset/dataset_upload_view.tsx:615]: I applied your suggestion. Please have a look.

**Review by philippotto — APPROVED** [review]
LGTM :+1:

**Pull request merged**